File: runtime/go/dispatcher.go
package runtime

import (
	"errors"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"slices"
	"sync"

	"github.com/Bitspark/nightseam/duplex/go"
)

// HandlerRegistry is the explicit registration capability used by generated
// bindings. Closing it releases its registrations, not its borrowed carrier.
type HandlerRegistry interface {
	bitwire.Wire
	Register([]string, bitwire.Receiver) (func(), error)
	Close(bitwire.Code, string) error
}

type dispatchRegistration struct {
	path     []string
	receiver bitwire.Receiver
}
type dispatchRoute struct {
	name   string
	prefix bool
}

// Dispatcher owns one endpoint attachment and an explicit exact/longest-prefix
// routing policy. It captures each request's traversal on the invocation its
// return capability carries, through the public vocabulary alone, and refuses a
// request whose return capability carries none rather than routing it with
// weaker detach and cancellation guarantees. An opaque wrapper is therefore as
// good as a native endpoint: the lifecycle travels with the unchanged return
// capability, and nothing here recognizes a concrete type.
type Dispatcher struct {
	root        bitwire.Endpoint
	ownEndpoint bool
	mu          sync.Mutex
	closed      bool
	detach      func()
	routes      map[dispatchRoute]*dispatchRegistration
}

// DispatcherOptions explicitly transfers closure authority for an endpoint the
// caller owns. Borrowed endpoints remain the default.
type DispatcherOptions struct{ OwnEndpoint bool }

func NewDispatcher(root bitwire.Endpoint, options ...DispatcherOptions) (*Dispatcher, error) {
	if root == nil {
		return nil, errors.New("dispatcher requires an endpoint")
	}
	d := &Dispatcher{root: root, routes: map[dispatchRoute]*dispatchRegistration{}}
	if len(options) > 0 {
		d.ownEndpoint = options[0].OwnEndpoint
	}
	detach, err := root.Receive(bitwire.Receiver{Message: d.deliver, Closed: func(code bitwire.Code, reason string) { _ = d.Close(code, reason) }})
	if err != nil {
		return nil, err
	}
	d.mu.Lock()
	closed := d.closed
	if !closed {
		d.detach = detach
	}
	d.mu.Unlock()
	if closed {
		detach()
		return nil, ErrClosed
	}
	return d, nil
}

func (d *Dispatcher) Send(path []string, message bitwire.Message) error {
	d.mu.Lock()
	closed := d.closed
	d.mu.Unlock()
	if closed {
		return ErrClosed
	}
	return d.root.Send(path, message)
}
func (d *Dispatcher) Register(path []string, receiver bitwire.Receiver) (func(), error) {
	return d.register(path, receiver, false)
}
func (d *Dispatcher) RegisterPrefix(path []string, receiver bitwire.Receiver) (func(), error) {
	return d.register(path, receiver, true)
}
func (d *Dispatcher) register(path []string, receiver bitwire.Receiver, prefix bool) (func(), error) {
	name, err := duplex.EncodePath(path)
	if err != nil {
		return nil, err
	}
	key := dispatchRoute{name, prefix}
	registration := &dispatchRegistration{path: slices.Clone(path), receiver: receiver}
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.closed {
		return nil, ErrClosed
	}
	if d.routes[key] != nil {
		return nil, duplex.ErrReceiverExists
	}
	d.routes[key] = registration
	return func() {
		d.mu.Lock()
		if d.routes[key] == registration {
			delete(d.routes, key)
		}
		d.mu.Unlock()
	}, nil
}
func (d *Dispatcher) match(path []string, name string) *dispatchRegistration {
	if exact := d.routes[dispatchRoute{name, false}]; exact != nil {
		return exact
	}
	var selected *dispatchRegistration
	for key, candidate := range d.routes {
		if key.prefix && len(candidate.path) <= len(path) && (selected == nil || len(candidate.path) > len(selected.path)) && slices.Equal(candidate.path, path[:len(candidate.path)]) {
			selected = candidate
		}
	}
	return selected
}
func (d *Dispatcher) deliver(path []string, message bitwire.Message) {
	name, err := duplex.EncodePath(path)
	if err != nil {
		return
	}
	// A control belongs to the traversal that captured it, never to the
	// registration in force now. Handing it to the invocation is what keeps a
	// detach or a rebind from retargeting an admitted request.
	if message.Frame.Kind == bitwire.ProfileCancel {
		_ = RelayInvocationControl(message)
		return
	}
	d.mu.Lock()
	var registration *dispatchRegistration
	if !d.closed {
		registration = d.match(path, name)
	}
	d.mu.Unlock()
	if registration == nil || registration.receiver.Message == nil {
		if message.Frame.Kind == bitwire.ProfileRequest {
			_ = sendWireResponse(message, nil, &PublicError{Code: "method_not_found", Message: "Unknown method"})
		}
		return
	}
	delivered := slices.Clone(path)
	if message.Frame.Kind != bitwire.ProfileRequest {
		registration.receiver.Message(delivered, message)
		return
	}
	capture, err := CaptureInvocation(message, func(control bitwire.Message) {
		registration.receiver.Message(slices.Clone(delivered), control)
	})
	if err != nil {
		// A bound reached is a refusal to try again at; a capability that
		// carries no lifecycle is a request this dispatcher cannot route with
		// the guarantees it advertises.
		refusal := &PublicError{Code: "invalid_message", Message: "Invocation requires the lifecycle its return capability carries"}
		if errors.Is(err, ErrInvocationLimit) {
			refusal = &PublicError{Code: "busy", Message: "Invocation participation limit reached"}
		}
		_ = sendWireResponse(message, nil, refusal)
		return
	}
	defer capture.Ready()
	registration.receiver.Message(delivered, message)
}

func (d *Dispatcher) Close(code bitwire.Code, reason string) error {
	d.mu.Lock()
	if d.closed {
		d.mu.Unlock()
		return nil
	}
	d.closed = true
	detach, routes := d.detach, d.routes
	d.detach, d.routes = nil, nil
	d.mu.Unlock()
	if detach != nil {
		detach()
	}
	for _, registration := range routes {
		if registration.receiver.Closed != nil {
			func() { defer func() { _ = recover() }(); registration.receiver.Closed(code, reason) }()
		}
	}
	if d.ownEndpoint {
		return d.root.Close(code, reason)
	}
	return nil
}

// Select returns a receiving view of this shared dispatcher. The view owns its
// prefix route and never acquires closure authority over the root endpoint.
func (d *Dispatcher) Select(path []string) *SelectedEndpoint {
	return &SelectedEndpoint{owner: d, prefix: slices.Clone(path)}
}

type SelectedEndpoint struct {
	owner      *Dispatcher
	prefix     []string
	mu         sync.Mutex
	closed     bool
	attachment *selectedAttachment
}
type selectedAttachment struct {
	receiver bitwire.Receiver
	detach   func()
}

func (s *SelectedEndpoint) Select(path []string) *SelectedEndpoint {
	return s.owner.Select(append(slices.Clone(s.prefix), path...))
}
func (s *SelectedEndpoint) Send(path []string, message bitwire.Message) error {
	s.mu.Lock()
	closed := s.closed
	s.mu.Unlock()
	if closed {
		return ErrClosed
	}
	return s.owner.Send(append(slices.Clone(s.prefix), path...), message)
}
func (s *SelectedEndpoint) Receive(receiver bitwire.Receiver) (func(), error) {
	attachment := &selectedAttachment{receiver: receiver}
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil, ErrClosed
	}
	if s.attachment != nil {
		s.mu.Unlock()
		return nil, duplex.ErrReceiverExists
	}
	s.attachment = attachment
	s.mu.Unlock()
	detach, err := s.owner.RegisterPrefix(s.prefix, bitwire.Receiver{
		Message: func(path []string, message bitwire.Message) {
			if receiver.Message != nil {
				receiver.Message(slices.Clone(path[len(s.prefix):]), message)
			} else if message.Frame.Kind == bitwire.ProfileRequest {
				_ = sendWireResponse(message, nil, &PublicError{Code: "method_not_found", Message: "Unknown method"})
			}
		},
		Closed: func(code bitwire.Code, reason string) { s.remove(attachment, true, code, reason) },
	})
	s.mu.Lock()
	active := s.attachment == attachment
	if err != nil && active {
		s.attachment = nil
	}
	if err == nil && active {
		attachment.detach = detach
	}
	s.mu.Unlock()
	if err != nil {
		return nil, err
	}
	if !active {
		detach()
		return nil, ErrClosed
	}
	return func() { s.remove(attachment, false, 0, "") }, nil
}
func (s *SelectedEndpoint) remove(attachment *selectedAttachment, tell bool, code bitwire.Code, reason string) {
	s.mu.Lock()
	if s.attachment != attachment {
		s.mu.Unlock()
		return
	}
	s.attachment = nil
	detach := attachment.detach
	s.mu.Unlock()
	if detach != nil {
		detach()
	}
	if tell && attachment.receiver.Closed != nil {
		attachment.receiver.Closed(code, reason)
	}
}
func (s *SelectedEndpoint) Close(code bitwire.Code, reason string) error {
	s.mu.Lock()
	s.closed = true
	attachment := s.attachment
	s.mu.Unlock()
	if attachment != nil {
		s.remove(attachment, true, code, reason)
	}
	return nil
}

File: runtime/ts/src/dispatcher.ts
import { encodePath, WireError } from '@nightseam/duplex';
import type { Endpoint, Message, Path, Receiver, Wire } from '@nightseam/duplex';
import { DuplexError } from './error.ts';
import { InvocationError, captureInvocation, relayInvocationControl } from './invocation.ts';
import { response } from './wire.ts';

/** Registration authority; close releases this registry, never its borrowed carrier. */
export interface HandlerRegistry extends Wire {
  register(path: Path, receiver: Receiver): () => void;
  close(code?: number, reason?: string): void;
}
interface Registration {
  path: Path;
  receiver: Receiver;
}

/** One attachment's options. */
export interface DispatcherOptions {
  /** Explicit closure authority over an endpoint owned by the caller. */
  ownEndpoint?: boolean;
}
/**
 * One attachment, with explicit exact-before-longest-prefix dispatch. Each
 * request's traversal is captured on the invocation its return capability
 * carries, through the public vocabulary alone; a request whose return
 * capability carries none is refused rather than routed with weaker detach and
 * cancellation guarantees. An opaque wrapper is therefore as good as a native
 * endpoint: the lifecycle travels with the unchanged return capability, and
 * nothing here recognizes a concrete type.
 */
export class WireDispatcher implements HandlerRegistry {
  private readonly exact = new Map<string, Registration>();
  private readonly prefixes = new Map<string, Registration>();
  private ended = false;
  private detach: (() => void) | undefined;
  private readonly root: Endpoint;
  private readonly ownEndpoint: boolean;

  constructor(root: Endpoint, options: DispatcherOptions = {}) {
    this.root = root;
    this.ownEndpoint = options.ownEndpoint ?? false;
    const detach = root.receive({
      message: (path, message) => this.deliver(path, message),
      closed: (code, reason) => this.close(code, reason),
    });
    if (this.ended) {
      detach();
      throw new WireError('closed');
    }
    this.detach = detach;
  }
  send(path: Path, message: Message): void {
    if (this.ended) throw new WireError('closed');
    this.root.send(path, message);
  }
  register(path: Path, receiver: Receiver): () => void {
    return this.install(path, receiver, this.exact);
  }
  registerPrefix(path: Path, receiver: Receiver): () => void {
    return this.install(path, receiver, this.prefixes);
  }
  private install(path: Path, receiver: Receiver, routes: Map<string, Registration>): () => void {
    const name = encodePath(path);
    if (this.ended) throw new WireError('closed');
    if (routes.has(name)) throw new WireError('receiver_exists');
    const registration = { path: [...path], receiver };
    routes.set(name, registration);
    return () => {
      if (routes.get(name) === registration) routes.delete(name);
    };
  }
  private match(path: Path, name: string): Registration | undefined {
    const exact = this.exact.get(name);
    if (exact) return exact;
    let selected: Registration | undefined;
    for (const candidate of this.prefixes.values()) {
      if (
        candidate.path.length <= path.length &&
        (!selected || candidate.path.length > selected.path.length) &&
        candidate.path.every((part, index) => part === path[index])
      )
        selected = candidate;
    }
    return selected;
  }
  private deliver(path: Path, message: Message): void | Promise<void> {
    const name = encodePath(path);
    // A control belongs to the traversal that captured it, never to the
    // registration in force now. Handing it to the invocation is what keeps a
    // detach or a rebind from retargeting an admitted request.
    if (message.frame.kind === 'cancel') {
      try {
        relayInvocationControl(message);
      } catch {
        /* A carrier with no lifecycle has nothing to route it to. */
      }
      return;
    }
    const registration = this.ended ? undefined : this.match(path, name);
    if (!registration?.receiver.message) {
      if (message.frame.kind === 'request')
        response(message, undefined, new DuplexError('method_not_found', 'Unknown method.'));
      return;
    }
    const delivered = [...path];
    if (message.frame.kind !== 'request') return registration.receiver.message(delivered, message);
    let capture;
    try {
      capture = captureInvocation(message, (control) => {
        void registration.receiver.message?.([...delivered], control);
      });
    } catch (error) {
      // A bound reached is a refusal to try again at; a capability that
      // carries no lifecycle is a request this dispatcher cannot route with
      // the guarantees it advertises.
      response(
        message,
        undefined,
        error instanceof InvocationError && error.code === 'limit'
          ? new DuplexError('busy', 'Invocation participation limit reached.')
          : new DuplexError('invalid_message', 'Invocation requires the lifecycle its return capability carries.'),
      );
      return;
    }
    let pending: void | Promise<void>;
    try {
      pending = registration.receiver.message(delivered, message);
    } catch (error) {
      capture.ready();
      throw error;
    }
    if (!pending) {
      capture.ready();
      return;
    }
    return pending.finally(() => capture.ready());
  }

  select(path: Path): SelectedEndpoint {
    return new SelectedEndpoint(this, [...path]);
  }
  close(code = 1000, reason = ''): void {
    if (this.ended) return;
    this.ended = true;
    const detach = this.detach;
    this.detach = undefined;
    const ending = [...this.exact.values(), ...this.prefixes.values()];
    this.exact.clear();
    this.prefixes.clear();
    detach?.();
    for (const registration of ending) {
      try {
        registration.receiver.closed?.(code, reason);
      } catch {
        /* Each owner receives its end. */
      }
    }
    if (this.ownEndpoint) this.root.close(code, reason);
  }
}
export function createDispatcher(endpoint: Endpoint, options: DispatcherOptions = {}): WireDispatcher {
  return new WireDispatcher(endpoint, options);
}

interface Attachment {
  receiver: Receiver;
  detach?: () => void;
}
/** A route owned by one dispatcher, with no borrowed-root closure authority. */
export class SelectedEndpoint implements Endpoint {
  private ended = false;
  private attachment: Attachment | undefined;
  private readonly owner: WireDispatcher;
  private readonly prefix: Path;
  constructor(owner: WireDispatcher, prefix: Path) {
    this.owner = owner;
    this.prefix = prefix;
  }
  select(path: Path): SelectedEndpoint {
    return this.owner.select([...this.prefix, ...path]);
  }
  send(path: Path, message: Message): void {
    if (this.ended) throw new WireError('closed');
    this.owner.send([...this.prefix, ...path], message);
  }
  receive(receiver: Receiver): () => void {
    if (this.ended) throw new WireError('closed');
    if (this.attachment) throw new WireError('receiver_exists');
    const attachment: Attachment = { receiver };
    this.attachment = attachment;
    try {
      const detach = this.owner.registerPrefix(this.prefix, {
        message: (path, message) => {
          if (receiver.message) return receiver.message(path.slice(this.prefix.length), message);
          if (message.frame.kind === 'request')
            response(message, undefined, new DuplexError('method_not_found', 'Unknown method.'));
        },
        closed: (code, reason) => this.remove(attachment, { code, reason }),
      });
      if (this.attachment !== attachment) {
        detach();
        throw new WireError('closed');
      }
      attachment.detach = detach;
    } catch (error) {
      if (this.attachment === attachment) this.attachment = undefined;
      throw error;
    }
    return () => this.remove(attachment);
  }
  private remove(attachment: Attachment, ending?: { code: number; reason: string }): void {
    if (this.attachment !== attachment) return;
    this.attachment = undefined;
    attachment.detach?.();
    if (ending) attachment.receiver.closed?.(ending.code, ending.reason);
  }
  close(code = 1000, reason = ''): void {
    this.ended = true;
    if (this.attachment) this.remove(this.attachment, { code, reason });
  }
}
